File: PracticeAssignment2Part1/Q4_Queue.java
package PracticeAssignment2Part1;

import java.util.ArrayList;
import java.util.List;

public class Q4_Queue {
     class Node{
        int data;
        Node next;

        public Node(int data){
            this.data = data;
            next = null;
        }
     }
     Node front;
     Node tail;

     public int peek(){
        if(front==null){
            return -1; // here i have put in -1 as throwing exception!
        }
        return front.data;
     }

     public void enqueue(int x){
        Node newNode = new Node(x);
        if(front==null){
            front = newNode;
        }else{
            tail.next = newNode;
            
        }
        tail = newNode;
        return;
     }
     public int dequeue(){
        if(front==null){
            return -1; //thorwing exception
        }
        Node deletedNode = front;
        Node newFront = front.next;
        if(newFront==null){
            tail = newFront;
        }
        front = newFront;
        return deletedNode.data;
     }
     public boolean isEmpty(){
        return front == null;
     }
}

File: homework2/saad_amin/q4_Queue.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class Queue:
    def __init__(self):
        self.head = None
        self.tail = None

    def peek(self):
        if self.head is None:
            return None
        return self.head.data
    
    def enqueue(self, x):
        new_node = Node(x)

        if self.head is None:
            self.head = self.tail = new_node
            return

        self.tail.next = new_node
        self.tail = new_node

    def dequeue(self):
        if self.head is None:
            return None

        val = self.head.data
        self.head = self.head.next

        if self.head is None:
            self.tail = None
        return val
    
    def isEmpty(self):
        return self.head is None